### PR TITLE
Protect encoding::decode_varint from overflow

### PR DESCRIPTION
**Pull request by ajguerrer**
Fixes the infamous #528.

### CODE REVIEW

File: src/encoding.rs [hunk 1]
```diff
@@ -167,7 +169,12 @@ where
         let byte = buf.get_u8();
         value |= u64::from(byte & 0x7F) << (count * 7);
         if byte <= 0x7F {
-            return Ok(value);
+            // check for u64::MAX overflow
+            if count == 9 && byte >= 2 {
```
Comment (ajguerrer) [inline src/encoding.rs:173]: Decided to add this check here in case moving it outside the first `if` caused some performance impact.

**Comment by LucioFranco** [comment]
Do you know if the C++ code protects against this? From what I can tell it does not.

**Comment by ajguerrer** [comment]
> Do you know if the C++ code protects against this? From what I can tell it does not.

@LucioFranco, your right! The C++ code will happily parse and overflow. I found this while studying the [Go code](https://github.com/protocolbuffers/protobuf-go/blob/master/encoding/protowire/wire.go#L358).

I understand that fixing this bug has debatable usefulness. It will not hurt my feelings at all if you decide not to merge this. I just thought it was interesting.

**Comment by LucioFranco** [comment]
I appreciate it! @ajguerrer do you mind updating the code with a reference to that go code, both the function doc and the line you changed? I think if we have that then I am fine to merge, I just wanted to check if there was precedence.

**Comment by LucioFranco** [comment]
Oh and also rebase against master

**Comment by ajguerrer** [comment]
@LucioFranco hows this look?

**Review by LucioFranco — APPROVED** [review]

**Comment by LucioFranco** [comment]
@ajguerrer there is one rustfmt error if you can fix that then we can merge

**Comment by ajguerrer** [comment]
@LucioFranco Once more!

**Pull request merged**